### DISCUSSION TITLE
Questions about PointNet++'s inference accuracy

### DISCUSSION THREAD

**Discussion by crea397** [pyg-team/pytorch_geometric#2758] [Q&A]
Hi,

I implemented the program with reference to examples/pointnet2_classification.py and used Google Colablatory's GPU / Ubuntu PC's GPU to learn the model.

When I save a model trained on a Colab / Ubuntu PC and load that model into Jetson Xavier NX for inference, the inference accuracy is much lower than when I do it on a Colab / Ubuntu PC.
<br>

There is no particular loss of accuracy between devices other than Jetson, for example, when saving a model trained on Colab, loading it on an Ubuntu PC, and performing inference.
There is also no loss of accuracy when saving models trained on Ubuntu PC and loading them in Colab.
<br>

When I save a model trained on another device (Colab, Ubuntu PC), load it into Jetson Xavier NX, and perform inference, I can't figure out why the accuracy drops.
<br>

How can I solve this problem?
Thanks!

---

**Enviroment**

$ uname -a
Linux JestonXavierNX 4.9.140-tegra #1 SMP PREEMPT Tue Oct 27 21:02:46 PDT 2020 aarch64 aarch64 aarch64 GNU/Linux

|Device  |Xavier NX  |Colab  |Ubuntu PC  |
|---|---|---|---|
|OS  |Ubuntu 18.04  |Ubuntu 18.04  |Ubuntu 20.04
|Python  |3.6.9  |3.7.10  |3.6.13  |
|torch  |1.7.0  |1.7.0  |1.7.0  |
|torch.version.cuda  |10.2  |10.2  |10.2  |
|torch-scatter  |2.0.5  |2.0.5  |2.0.7  |
|torch-sparse  |0.6.8  |0.6.8  |0.6.10  |
|torch-cluster  |1.5.8  |1.5.8  |1.5.9  |
|torch-spline-conv  |1.2.0  |1.2.0  |1.2.1  |
|torch-geometric  |1.6.1  |1.6.1  |1.7.1  |

---

- The same test data is used.
- Saving the models trained by Jetson Xavier NX and loading them into Jetson Xavier NX for inference works fine.
- The following code is used to save and load the model.

```Python
# save
torch.save(model.state_dict(), './xxx.pth')

# load
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net()
model.load_state_dict(torch.load('./xxx.pth'))
model.to(device)
```

**Comment by rusty1s** [comment]
Do you know if this is only the case for the `PointNet++` model? Do you know which model operation causes the divergence?

**Reply by crea397** [reply]
> Do you know if this is only the case for the PointNet++ model?

I used PyTorch to create a model for classifying handwritten numbers in MNIST.
I used Ubuntu PC's GPU to train the model. 
I saved the model trained on my Ubuntu PC, loaded it into the Jetson Xavier NX and inferred it, and got the same inference accuracy as I did on my Ubuntu PC.
Therefore, it may be said that the above case will not happen except for PointNet++.

<br>

> Do you know which model operation causes the divergence?

I checked the contents of variables when doing inference with PointNet++ on Jetson and Ubuntu PC.
The source code for doing inference with PointNet++ is the same for both Jetson and Ubuntu PCs.
Unfortunately, I did not see any difference in the dtype or device of PyTorch.

**Reply by rusty1s** [reply]
Thanks for digging into this. It would be great to check if this only happens for `PointNet++` or other PyG models as well, such as a simple `GCN` in `examples/gcn.py`.

Is there any chance you can debug the intermediate outputs of PointNet++?

**Reply by crea397** [reply]
> Thanks for digging into this. It would be great to check if this only happens for PointNet++ or other PyG models as well, such as a simple GCN in `examples/gcn.py`.

Thank you for your suggestion!

I tried the same thing with `examples/gcn.py`, but the accuracy did not decrease with Jetson Xavier NX, and the accuracy was the same on Ubuntu PC and Jetson Xavier NX.

<br>

> Is there any chance you can debug the intermediate outputs of PointNet++?

I have displayed the intermediate output of PointNet++ (Jetson Xavier NX, Ubuntu PC) for a single test data.
In the following screenshot, the variable names, and the numbers in front of the variable names, correspond to the lines in `examples/pointnet2_classification.py`.

Comparing the intermediate output, the difference seems to be in the `row` and `col` of the `SAModule`.
The output will follow the following format (variable.device, variable.dtype, len(variable)).

<img width="357" alt="Jetson Xavier NX" src="https://user-images.githubusercontent.com/63713337/122743299-d6406e80-d2c1-11eb-9ebe-80dad6027ab5.png">      <img width="360" alt="Ubuntu PC" src="https://user-images.githubusercontent.com/63713337/122743321-db9db900-d2c1-11eb-83ed-e46b7c9a488a.png">

**Reply by rusty1s** [reply]
Yes, it looks like that `radius` will produce slight different graphs. Although the `radius` call is non-deterministic, the shape of `edge_index` should match nonetheless. You may want to confirm that `radius` produces similar graphs on both machines. In addition, it might be a good idea to replace `radius` with `knn` to see if that fixes your issues.

**Reply by crea397** [reply]
How can I confirm that `radius` produces a similar graph on both machines?
At first, I checked the contents of the `row` and `col` generated by `radius` by printing them, and found that the contents were similar.

I tried replacing `radius` with `knn`, but unfortunately the issue is not fixed.
As with `radius`, the accuracy was reduced.

**Reply by rusty1s** [reply]
Given that `max_num_neighbors` is set to a high value in which no potential neighbors inside query balls are dropped, the outputs should be equal. The output for `knn` should be equal as well.

**Reply by crea397** [reply]
Thank you very much for your advice.
I have confirmed that when I replace `radius` with `knn`, the length of `row` and `col` output by `knn` are the same whether `k=32` or `k=64`. However, the accuracy is reduced in Jetson.

When using `radius`, I have confirmed that the lengths of `row` and `col` output by `radius` do not match, even if I set the value of `max_num_neighbors` to a high value. Also, the accuracy is reduced in Jetson.

In the screenshot above, when `max_num_neighbors=64`, the length of `row` and `col` output by `radius` is shorter in Jetson. As I increase the value of `max_num_neighbors` on Jetson, those values approach and eventually exceed the length of the `row` and `col` output by `radius` on Ubuntu PC.

I also tried increasing the value of `max_num_neighbors` by the same amount on both Jetson and Ubuntu PC, but the length of `row` and `col` output by `radius` on both did not match.

The values of `max_num_neighbors` I tried are 128, 256, 512, and 1024.

**Reply by rusty1s** [reply]
Thank you very much. It is really weird that both `radius` and `knn` perform differently on Jetson. Are you sure that the input point cloud is the same in both cases? Is there a possibility that points are scaled differently?
I'm also wondering if you can confirm whether `radius` yields the correct result on Jetson and Ubuntu. What happens if you input a simple toy example with say 100 points?

**Reply by crea397** [reply]
> Are you sure that the input point cloud is the same in both cases? Is there a possibility that points are scaled differently?

I think the same input point cloud is used.
For the input point cloud, I am applying `T.NormalizeScale()` on both machines.
Can this be a possibility that the input point cloud is different or the point scaling is different?
Just to be sure, I also tried not applying `T.NormalizeScale()` on both machines. However, the accuracy was reduced by Jetson as well.


> What happens if you input a simple toy example with say 100 points?

I used the data of a rabbit (113 points) and a chair (111 points)  as input point clouds and checked the output of `radius`.
In this case, I did not apply `T.NormalizeScale()` on both machines. Also, I used a value of 64 for `max_num_neighbors`.

The result is shown in the screenshot below.
The output results of `radius` on both machines did not match.

## rabbit

![r1](https://user-images.githubusercontent.com/63713337/123414671-27689f00-d5ef-11eb-9200-2c4693e16e69.png)![r2](https://user-images.githubusercontent.com/63713337/123414683-2afc2600-d5ef-11eb-9e8a-745088a67c42.png)

## chair

![c1](https://user-images.githubusercontent.com/63713337/123414715-351e2480-d5ef-11eb-8139-f4cb5ab8c5f9.png)![c2](https://user-images.githubusercontent.com/63713337/123414728-37807e80-d5ef-11eb-9b85-cfb35e516cc5.png)

**Reply by rusty1s** [reply]
Ok, thanks a lot for digging into this so closely. At this point, we should probably move this discussion over to `pytorch-cluster`. Feel free to create an issue there. Can you run the test-suite of pytorch-cluster in Jetson and post the output log?

**Reply by crea397** [reply]
Thank you very much for your suggestion. I will create a new issue in the `pytorch-cluster` repository and post the output log.

Sorry, but please let me know before I do that. What is the test-suite for `pytorch-cluster`?

Do I just run `python setup.py test` on https://github.com/rusty1s/pytorch_cluster and post the output?
For example, or do I just run `pytorch_cluster/test/test_radius.py` and post the output for each variable as before?

**Reply by rusty1s** [reply]
You can run `python setup.py test` and let me know the error log of tests that fail.
